### PR TITLE
Refactor dev

### PR DESCRIPTION
**Pull request by hillol-pal**
pull request for refactoring a Simple Crud Reminder Service as a microservice app

### CODE REVIEW

File: microservice/challenge-1/ReminderAppl/Docker-Compose/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    image: mysql:latest
+    container_name: mysql-local
+    environment:
+      - MYSQL_ROOT_PASSWORD=default
+      - MYSQL_DATABASE=reminderDB
+      - MYSQL_USER=secret
+      - MYSQL_PASSWORD=secretpassword
+    restart: always
+    volumes:
+      - C:\\gitrepos\\mysql-vol:/var/lib/mysql
```
Comment (albertaouad) [inline microservice/challenge-1/ReminderAppl/Docker-Compose/docker-compose.yml:25]: windows specific

File: microservice/challenge-1/ReminderAppl/Docker-Compose/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+version: "3.3"
+
+services:
+
+  reminder-app-local:
+    image: reminder-app:20.0
```
Comment (albertaouad) [inline microservice/challenge-1/ReminderAppl/Docker-Compose/docker-compose.yml:6]: where's the build step, this image does not exist on my local.

Comment (albertaouad) [inline microservice/challenge-1/ReminderAppl/Docker-Compose/docker-compose.yml:6]:
```suggestion
    image: reminder-app:20.0
    build:
      context: ../
      dockerfile: Dockerfile
```

File: microservice/challenge-1/ReminderAppl/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,15 @@
+FROM openjdk:8-jdk-alpine
```
Comment (albertaouad) [inline microservice/challenge-1/ReminderAppl/Dockerfile:1]: Use a multistage dockerfile to build the artifact in one layer and serve it in another

File: microservice/challenge-1/ReminderAppl/README.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
+Context:
```
Comment (albertaouad) [inline microservice/challenge-1/ReminderAppl/README.md:1]: Missing markdown formatting.
Missing steps to build locally, step to run locally, steps to run with docker and docker-compose.
Steps to package and deploy helm chart.

File: microservice/challenge-1/ReminderAppl/charts/values.yaml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+podAnnotations: {}
+
+podSecurityContext: {}
+
+securityContext: {}
+
+
+service:
+  type: NodePort
+  port: 33706
```
Comment (albertaouad) [inline microservice/challenge-1/ReminderAppl/charts/values.yaml:35]: Can I ask you why you're using a NodePort and not a ClusterIP ?

**Comment by albertaouad** [comment]
@hillol-pal  thank you for your submission, I left you a couple comments as I did a quick skim through your code.
If I may ask you please add steps in Readme to run the app locally and from docker-compose for me to validate that your code works in the way you expect us to run it.

On top of that,  I'm interested to know what would be the second microservice to be able to state that you completely migrated the JAxRSREminder app fully?
Also, If you were to handle User Authentication and Authorization, how would you tackle that (no need for implementation, high level design)?
And last how would you handle scalability, for instance your app now has 1M+ active users that are all using it on a daily basis, what would you look at optimizing to make sure your users continue to have a smooth experience?

File: microservice/challenge-1/ReminderAppl/Docker-Compose/docker-compose.yml (hunk 1 above)
Comment (hillol-pal) [reply microservice/challenge-1/ReminderAppl/Docker-Compose/docker-compose.yml]: Removed and used internal volume.

File: microservice/challenge-1/ReminderAppl/Dockerfile (hunk 3 above)
Comment (hillol-pal) [reply microservice/challenge-1/ReminderAppl/Dockerfile]: Multi-stage docker build applied

File: microservice/challenge-1/ReminderAppl/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+RUN apk --no-cache add curl
+COPY --from=basebuilder /app/src/target/reminder-app*.jar /app/app.jar
+ENTRYPOINT ["java","-jar","/app/app.jar","--spring.profiles.active=dev"]
+
+FROM openjdk:8-jdk-alpine as prod
+
+RUN apk update && apk upgrade && apk add --no-cache bash
+RUN apk --no-cache add curl
+COPY --from=basebuilder /app/src/target/reminder-app*.jar /app/app.jar
+ENTRYPOINT ["java","-jar","/app/app.jar","--spring.profiles.active=prod"]
```
Comment (albertaouad) [inline microservice/challenge-1/ReminderAppl/Dockerfile:20]: That's not why multistage builds are used, changing profiles should be done by env vars.

File: microservice/challenge-1/ReminderAppl/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,20 @@
+FROM openjdk:8-jdk-alpine as basebuilder
+
+RUN mkdir -p /app/src
+COPY . /app/src
+WORKDIR /app/src
+RUN ./mvnw clean package
```
Comment (albertaouad) [inline microservice/challenge-1/ReminderAppl/Dockerfile:6]: this step fails to run locally, mvnw is not part of the binaries I have locally

File: microservice/challenge-1/ReminderAppl/Docker-Compose/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+services:
+
+  reminder-app-local:
+    build:
+      context: ../
+      dockerfile: Dockerfile
+      target: dev
+    container_name: reminder-app-local
+    env_file: ./.env
```
Comment (albertaouad) [inline microservice/challenge-1/ReminderAppl/Docker-Compose/docker-compose.yml:11]: the checked in env file is call .dev.env not .env

File: microservice/challenge-1/ReminderAppl/Docker-Compose/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+      - "mysql-local"
+    networks:
+      - reminderapp-network
+    ports:
+      - "8080:8080"
+
+  mysql-local:
+    image: mysql:latest
+    container_name: mysql-local
+    env_file: ./.env
```
Comment (albertaouad) [inline microservice/challenge-1/ReminderAppl/Docker-Compose/docker-compose.yml:27]: the checked in env file is call .dev.env not .env

File: microservice/challenge-1/ReminderAppl/Dockerfile (hunk 7 above)
Comment (hillol-pal) [reply microservice/challenge-1/ReminderAppl/Dockerfile:6]: Added the maven wrapper specific files

File: microservice/challenge-1/ReminderAppl/Docker-Compose/docker-compose.yml (hunk 8 above)
Comment (hillol-pal) [reply microservice/challenge-1/ReminderAppl/Docker-Compose/docker-compose.yml]: Updated

**Review by lucasledesma — COMMENTED** [review]
When running docker-compose up I got this error:

image

**Review by lucasledesma — COMMENTED** [review]
I can't start the app with docker compose, getting exceptions...

image

image

File: microservice/challenge-1/ReminderAppl/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,14 @@
+FROM openjdk:8-jdk-alpine as basebuilder
```
Comment (alygueye) [inline microservice/challenge-1/ReminderAppl/Dockerfile:1]: tried to build the image local and got permission error on `./mvnw`, is it possible to update the multi-stage build using a maven image instead?

image

**Review by alygueye — CHANGES_REQUESTED** [review]
@hillol-pal Thanks for the PR, got few error running the docker-compose

File: microservice/challenge-1/ReminderAppl/Dockerfile (hunk 10 above)
Comment (albertaouad) [reply microservice/challenge-1/ReminderAppl/Dockerfile:1]: you need chmod +x ./mvnw locally.

Comment (hillol-pal) [reply microservice/challenge-1/ReminderAppl/Dockerfile:1]: I added RUN chmod +x ./mvnw in the Dockerfile. Also tried with a maven image maven:3.5.2-jdk-8-alpine  and it worked for me locally.

**Comment by hillol-pal** [comment]
> I can't start the app with docker compose, getting exceptions...
> 
> image
> 
> image

@lucasledesma : This one looks like image specific issue and it is related to permissions created during the DB creation. I looked through the dockerhub page and Github issues specific to this one and suggestion was to add another environment variable MYSQL_ROOT_HOST=%. Can you please retry?

**Pull request closed without merging**